juju/juju
# [JUJU-1723] Implement support for k8s secrets backend

Add support for a k8s secrets backend.
You'll need to enable 2 feature flags - secrets and secrets-stores

On a k8s cloud, with the secrets-stores flag enabled, the secret content will be saved as k8s secrets.

## Checklist


## QA steps

```
export JUJU_DEV_FEATURE_FLAGS=secrets,secrets-stores

juju bootstrap microk8s test
Creating Juju controller "test" on microk8s/localhost
Bootstrap to Kubernetes cluster identified as microk8s/localhost
Creating k8s resources for controller "controller-test"
Starting controller pod
Bootstrap agent now started
Contacting Juju controller at 10.152.183.148 to verify accessibility...

Bootstrap complete, controller "test" is now available in namespace "controller-test"

Now you can run
        juju add-model <model-name>
to create a new model to deploy k8s workloads.

juju switch controller
test (controller) -> test:admin/controller

juju exec --unit controller/0 "secret-add foo=bar"
secret:cc7fa0orsjec76phn6d0

juju exec --unit controller/0 "secret-get secret:cc7fa0orsjec76phn6d0"
foo: bar

juju exec --unit controller/0 "secret-update secret:cc7fa0orsjec76phn6d0 foo=bar2"

juju exec --unit controller/0 "secret-get secret:cc7fa0orsjec76phn6d0 --update"
foo: bar2

juju show-secret secret:cc7fa0orsjec76phn6d0 --reveal
cc7fa0orsjec76phn6d0:
  revision: 2
  owner: controller
  created: 2022-08-31T05:43:32Z
  updated: 2022-08-31T05:44:31Z
  content:
    foo: bar2

mkubectl -n controller-test get secrets
NAME                            TYPE                                  DATA   AGE
controller-proxy                kubernetes.io/service-account-token   3      2m37s
controller-secret               Opaque                                2      2m36s
controller-application-config   Opaque                                1      2m34s
cc7fa0orsjec76phn6d0-1          Opaque                                1      107s
cc7fa0orsjec76phn6d0-2          Opaque                                1      47s

mkubectl -n controller-test get -o json secret/cc7fa0orsjec76phn6d0-1 | jq .data
{
  "foo": "WW1GeQ=="
}

mkubectl -n controller-test get -o json secret/cc7fa0orsjec76phn6d0-2 | jq .data
{
  "foo": "WW1GeU1nPT0="
}

juju exec --unit controller/0 "secret-remove secret:cc7fa0orsjec76phn6d0"

mkubectl -n controller-test get secrets
NAME                            TYPE                                  DATA   AGE
controller-proxy                kubernetes.io/service-account-token   3      2m41s
controller-secret               Opaque                                2      2m40s
controller-application-config   Opaque                                1      2m38s
```
Do the same on an LXD controller to make sure the Juju backend works for a non-k8s cloud.
